ES2019 Object.fromEntries uses PropertyKeys as mapped key type

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.5.0-dev.20190514

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
fromEntries
**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
type T = {}

const a: { [key: string]: T } = { "key": {} }

const b: [string, T][] = Object.entries(a)

const c : {[key: number]: T; [key: string]: T } = Object.fromEntries(b)

const d : { [key: string]: T } = Object.fromEntries(b)

```

**Expected behavior:**

~Compilation should fail because `number` is not a key of `a`.~

_edit:_ The return type of `Object.fromEntries(b)` should be inferred as `d`: `{ [key: string]: T } ` .


**Actual behavior:**

_edit:_ The return type of `Object.fromEntries(b)` is `{[key: number]: T; [key: string]: T }` Adding an erroneous `[key: number]`.


`Object.fromEntries` has type:
```ts
 fromEntries<T = any>(entries: Iterable<readonly [PropertyKey, T]>): { [k in PropertyKey]: T };
```

However if I make a second type parameter for fromEntries which extends PropertyKey and use that for the key type in forEntries:
```ts
 fromEntries<K extends PropertyKey, T = any>(entries: Iterable<readonly [K, T]>): { [k in K]: T };
```

I'm willing to make a PR for this if this sounds like a reasonable approach.

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->

Apologies, but I could not use `es2019` in the playground. 

**Related Issues:** <!-- Did you find other bugs that looked similar? -->

https://github.com/microsoft/TypeScript/pull/30934

https://github.com/microsoft/TypeScript/issues/30933 https://github.com/microsoft/TypeScript/issues/25999

Got the same issue randomly, not sure what triggered it as things were rendering fine. But using a target of either `esnext` or `es2019` will cause this compilation issue. Also confirmed this by updating to `typescript@next` and running the same code.

Replication details are as follows, I did this in my `/tmp` directory and you can pretty much copy and paste these steps to reproduce.

**tsconfig.json**
```json
{
  "compilerOptions": {
    "target": "es2019",
    "keyofStringsOnly": true,
  },
}
```

Running `tsc` will complain that it has nothing to do, so  `touch index.ts`.
Then running the `tsc` command will result in the following error:
```
*chop*/typescript/lib/lib.es2019.object.d.ts:28:81 - error TS2322: Type 'string | number | symbol' is not assignable to type 'string'.
  Type 'number' is not assignable to type 'string'.

28     fromEntries<T = any>(entries: Iterable<readonly [PropertyKey, T]>): { [k in PropertyKey]: T };
                                                                                   ~~~~~~~~~~~

Found 1 error.
```

This is specifically part of `es2019` (currently `esnext` aliases this I believe).
The fix, for now, is to `"target": "es2018"` instead in your `tsconfig.json`. 

```json
{
  "compilerOptions": {
    "target": "es2018",
    "keyofStringsOnly": true,
  },
}
```

I changed `target` to `"target": "es2018",` but I am still getting the following error:

```ts
node_modules/typescript/lib/lib.es2019.object.d.ts:28:81 - error TS2322: Type 'string | number | symbol' is not assignable to type 'string'.
  Type 'number' is not assignable to type 'string'.

28     fromEntries<T = any>(entries: Iterable<readonly [PropertyKey, T]>): { [k in PropertyKey]: T };
```

Here is my `tsconfig.json`


```json
{
  "compilerOptions": {
    "target": "es2018",
    "moduleResolution": "node",
    "module": "commonjs",
    "sourceMap": true,
    "rootDir": "src",
    "outDir": "dist",
    "keyofStringsOnly": true,
    "lib": ["esnext", "dom"],
    "esModuleInterop": true,
  },
  "exclude": ["node_modules", "**/*.spec.ts"]
}
```

The only thing that works is removing `"keyofStringsOnly": true,`.

I am not sure why this is still happening. Has anyone else experienced this/ any suggestions on a fix?

--- a/src/lib/es2019.object.d.ts
+++ b/src/lib/es2019.object.d.ts
@@ -5,7 +5,7 @@ interface ObjectConstructor {
      * Returns an object created by key-value entries for properties and methods
      * @param entries An iterable object that contains key-value entries for properties and methods.
      */
-    fromEntries<T = any>(entries: Iterable<readonly [PropertyKey, T]>): { [k in PropertyKey]: T };
+    fromEntries<T = any>(entries: Iterable<readonly [PropertyKey, T]>): { [k: string]: T };
 
     /**
      * Returns an object created by key-value entries for properties and methods

--- a/tests/baselines/reference/objectFromEntries.types
+++ b/tests/baselines/reference/objectFromEntries.types
@@ -1,10 +1,10 @@
 === tests/cases/compiler/objectFromEntries.ts ===
 const o = Object.fromEntries([['a', 1], ['b', 2], ['c', 3]]);
->o : { [x: string]: number; [x: number]: number; }
->Object.fromEntries([['a', 1], ['b', 2], ['c', 3]]) : { [x: string]: number; [x: number]: number; }
->Object.fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [x: string]: T; [x: number]: T; }; (entries: Iterable<readonly any[]>): any; }
+>o : { [k: string]: number; }
+>Object.fromEntries([['a', 1], ['b', 2], ['c', 3]]) : { [k: string]: number; }
+>Object.fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [k: string]: T; }; (entries: Iterable<readonly any[]>): any; }
 >Object : ObjectConstructor
->fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [x: string]: T; [x: number]: T; }; (entries: Iterable<readonly any[]>): any; }
+>fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [k: string]: T; }; (entries: Iterable<readonly any[]>): any; }
 >[['a', 1], ['b', 2], ['c', 3]] : [string, number][]
 >['a', 1] : [string, number]
 >'a' : "a"
@@ -17,20 +17,20 @@ const o = Object.fromEntries([['a', 1], ['b', 2], ['c', 3]]);
 >3 : 3
 
 const o2 = Object.fromEntries(new URLSearchParams());
->o2 : { [x: string]: string; [x: number]: string; }
->Object.fromEntries(new URLSearchParams()) : { [x: string]: string; [x: number]: string; }
->Object.fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [x: string]: T; [x: number]: T; }; (entries: Iterable<readonly any[]>): any; }
+>o2 : { [k: string]: string; }
+>Object.fromEntries(new URLSearchParams()) : { [k: string]: string; }
+>Object.fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [k: string]: T; }; (entries: Iterable<readonly any[]>): any; }
 >Object : ObjectConstructor
->fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [x: string]: T; [x: number]: T; }; (entries: Iterable<readonly any[]>): any; }
+>fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [k: string]: T; }; (entries: Iterable<readonly any[]>): any; }
 >new URLSearchParams() : URLSearchParams
 >URLSearchParams : { new (init?: string | URLSearchParams | string[][] | Record<string, string>): URLSearchParams; prototype: URLSearchParams; toString(): string; }
 
 const o3 = Object.fromEntries(new Map([[Symbol("key"), "value"]]));
->o3 : { [x: string]: string; [x: number]: string; }
->Object.fromEntries(new Map([[Symbol("key"), "value"]])) : { [x: string]: string; [x: number]: string; }
->Object.fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [x: string]: T; [x: number]: T; }; (entries: Iterable<readonly any[]>): any; }
+>o3 : { [k: string]: string; }
+>Object.fromEntries(new Map([[Symbol("key"), "value"]])) : { [k: string]: string; }
+>Object.fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [k: string]: T; }; (entries: Iterable<readonly any[]>): any; }
 >Object : ObjectConstructor
->fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [x: string]: T; [x: number]: T; }; (entries: Iterable<readonly any[]>): any; }
+>fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [k: string]: T; }; (entries: Iterable<readonly any[]>): any; }
 >new Map([[Symbol("key"), "value"]]) : Map<symbol, string>
 >Map : MapConstructor
 >[[Symbol("key"), "value"]] : [symbol, string][]
@@ -60,9 +60,9 @@ const frozenArray = Object.freeze([['a', 1], ['b', 2], ['c', 3]]);
 const o4 = Object.fromEntries(frozenArray);
 >o4 : any
 >Object.fromEntries(frozenArray) : any
->Object.fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [x: string]: T; [x: number]: T; }; (entries: Iterable<readonly any[]>): any; }
+>Object.fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [k: string]: T; }; (entries: Iterable<readonly any[]>): any; }
 >Object : ObjectConstructor
->fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [x: string]: T; [x: number]: T; }; (entries: Iterable<readonly any[]>): any; }
+>fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [k: string]: T; }; (entries: Iterable<readonly any[]>): any; }
 >frozenArray : readonly (string | number)[][]
 
 const frozenArray2: readonly [string, number][] = Object.freeze([['a', 1], ['b', 2], ['c', 3]]);
@@ -83,10 +83,10 @@ const frozenArray2: readonly [string, number][] = Object.freeze([['a', 1], ['b',
 >3 : 3
 
 const o5 = Object.fromEntries(frozenArray2);
->o5 : { [x: string]: number; [x: number]: number; }
->Object.fromEntries(frozenArray2) : { [x: string]: number; [x: number]: number; }
->Object.fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [x: string]: T; [x: number]: T; }; (entries: Iterable<readonly any[]>): any; }
+>o5 : { [k: string]: number; }
+>Object.fromEntries(frozenArray2) : { [k: string]: number; }
+>Object.fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [k: string]: T; }; (entries: Iterable<readonly any[]>): any; }
 >Object : ObjectConstructor
->fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [x: string]: T; [x: number]: T; }; (entries: Iterable<readonly any[]>): any; }
+>fromEntries : { <T = any>(entries: Iterable<readonly [string | number | symbol, T]>): { [k: string]: T; }; (entries: Iterable<readonly any[]>): any; }
 >frozenArray2 : readonly [string, number][]
 

